pkg/flow/internal/controller/component_node_manager.go

```diff
@@ -0,0 +1,258 @@
+	m.additionalDeclareContents = additionalDeclareContents
+	m.customComponentDependencies = make(map[string][]CustomComponentDependency)
+	m.importNodes = map[string]*ImportConfigNode{}
+	m.declareNodes = map[string]*DeclareNode{}
+}
+
+// CreateComponentNode creates a new builtin component or a new custom component.
+func (m *ComponentNodeManager) CreateComponentNode(componentName string, block *ast.BlockStmt) (ComponentNode, error) {
+	if m.isCustomComponent(componentName) {
+		return NewCustomComponentNode(m.globals, block, m.GetCustomComponentConfig), nil
+	} else {
+		registration, exists := m.componentReg.Get(componentName)
+		if !exists {
+			return nil, fmt.Errorf("unrecognized component name %q", componentName)
+		}
+		return NewBuiltinComponentNode(m.globals, registration, block), nil
+	}
+}
+
+// isCustomComponent searches for a declare corresponding to the given component name.
+func (m *ComponentNodeManager) isCustomComponent(componentName string) bool {
+	m.mut.Lock()
+	defer m.mut.Unlock()
+	namespace := strings.Split(componentName, ".")[0]
+	_, declareExists := m.declareNodes[namespace]
+	_, importExists := m.importNodes[namespace]
+	_, additionalDeclareContentExists := m.additionalDeclareContents[componentName]
+
+	return declareExists || importExists || additionalDeclareContentExists
+}
+
+// FindLocalDeclareNode returns the declareNode matching the declareLabel of the provided CustomComponentNode if present.
+func (m *ComponentNodeManager) FindLocalDeclareNode(cc *CustomComponentNode) (*DeclareNode, bool) {
+	m.mut.Lock()
+	defer m.mut.Unlock()
+	declareNode, exist := m.declareNodes[cc.declareLabel]
+	return declareNode, exist
+}
+
+// FindImportedConfigNode returns the importNode matching the importLabel of the provided CustomComponentNode if present.
+func (m *ComponentNodeManager) FindImportedConfigNode(cc *CustomComponentNode) (*ImportConfigNode, bool) {
+	m.mut.Lock()
+	defer m.mut.Unlock()
+	importNode, exist := m.importNodes[cc.importLabel]
+	return importNode, exist
+}
+
+// CustomComponentConfig represents the config needed by a custom component to load.
+type CustomComponentConfig struct {
+	declareContent            string            // represents the corresponding declare as plain string
+	additionalDeclareContents map[string]string // represents the additional declares that might be needed by the component to build custom components
+}
+
+// NewCustomComponentConfig creates a new CustomComponentConfig.
+func NewCustomComponentConfig(declareContent string, additionalDeclareContents map[string]string) *CustomComponentConfig {
+	return &CustomComponentConfig{
+		declareContent:            declareContent,
+		additionalDeclareContents: additionalDeclareContents,
+	}
+}
+
+// GetCustomComponentConfig returns the custom component config for a given custom component or an error if not found.
+func (m *ComponentNodeManager) GetCustomComponentConfig(cc *CustomComponentNode) (*CustomComponentConfig, error) {
```

Comment:
It feels like we're introducing two different ways within the code to look up components now:

1. ComponentRegistry interface
2. The GetCustomComponentConfig function passed around to nodes. 

Is there a way for us to roll up the functionality for looking up all types of components into ComponentRegistry?

Reply:
You might find these two thought experiments useful:

1. What might it look like if builtin components and custom components were more aligned in implementation? 

2. What might an API look which has a lookup method that might return different component types? 

These are two distinct thought experiments that would lead to two different solutions for solving the problem I'd like to be investigated.

Reply:
Sorry I'm not sure I'm following. Even if they would be more aligned in implementation, they are by nature still very different. One works as a container for the other one

Reply:
To rephrase the first thought experiment: what might it look like for custom components to be represented by a component.Registration, allowing the caller to use them without knowledge of whether they're backed by a builtin or custom component?

Reply:
That sounds like the old modules implementation except that instead of the content of the module being loaded by the module it would be available inside of the component registry. The Build function would look the content up the scope and return the custom component ready to run?